src/iam/certhandler/modules/pkcs11.cpp

```diff
@@ -0,0 +1,986 @@
+struct SearchObject {
+    /**
+     * Object type.
+     */
+    Optional<pkcs11::ObjectClass> mType;
+    /**
+     * Object handle.
+     */
+    pkcs11::ObjectHandle mHandle;
+    /**
+     * Object label.
+     */
+    StaticString<pkcs11::cLabelLen> mLabel;
+    /**
+     * Key identifier for public/private key pair.
+     */
+    uuid::UUID mID;
+};
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+PKCS11Module::PKCS11Module(const String& certType, const ModuleConfig& config)
+    : mCertType(certType)
+    , mConfig(config)
+{
+}
+
+Error PKCS11Module::Init(
+    pkcs11::LibraryContext& pkcs11, crypto::x509::ProviderItf& x509Provider, uuid::UUIDManagerItf& uuidManager)
+{
+    mPKCS11 = &pkcs11;
+    mX509Provider = &x509Provider;
+    mUUIDManager = &uuidManager;
+
+    Error err = Error::Enum::eNone;
+    Tie(cTeeClientUUIDNs, err) = mUUIDManager->StringToUUID("58AC9CA0-2086-4683-A1B8-EC4BC08E01B6");
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    err = os::GetEnv(cEnvLoginType, mTeeLoginType);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (!mConfig.mUserPINPath.IsEmpty() && mTeeLoginType.IsEmpty()) {
+        return AOS_ERROR_WRAP(Error::Enum::eInvalidArgument);
+    }
+
+    err = pkcs11.InitLibrary(mConfig.mLibrary);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    mPKCS11 = &pkcs11;
+
+    mConfig.mTokenLabel = GetTokenLabel();
+
+    Tie(mSlotID, err) = GetSlotID();
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    bool isOwned = false;
+
+    Tie(isOwned, err) = IsOwned();
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    if (isOwned) {
+        err = PrintInfo(mSlotID);
+    } else {
+        LOG_DBG() << "No owned token found";
+    }
+
+    return err;
+}
+
+Error PKCS11Module::SetOwner(const String& password)
+{
+    Error err = Error::Enum::eNone;
+
+    Tie(mSlotID, err) = GetSlotID();
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    mPendingKeys.Clear();
+
+    if (!mTeeLoginType.IsEmpty()) {
+        StaticString<pkcs11::cPINLength> userPIN;
+
+        err = GetTeeUserPIN(mTeeLoginType, mConfig.mUID, mConfig.mGID, userPIN);
+        if (!err.IsNone()) {
+            return err;
+        }
+    } else {
+        StaticString<pkcs11::cPINLength> userPIN;
+
+        err = GetUserPin(userPIN);
+        if (!err.IsNone()) {
+            err = pkcs11::GenPIN(userPIN);
+            if (!err.IsNone()) {
+                return err;
+            }
+
+            err = FS::WriteFile(mConfig.mUserPINPath, userPIN, 0600);
+            if (!err.IsNone()) {
+                return err;
+            }
+        }
+    }
+
+    LOG_DBG() << "Init token. SlotID=" << mSlotID << ", label=" << mConfig.mTokenLabel;
```

Comment:
Let's have the following log format: `message: var1 = val1, var2 = var2`

`Init token: slotID = ..., label = ...`